### ISSUE TITLE
CTRL Modifier is sent later than subsequent Keypresses to the target system

### ISSUE THREAD

**Issue by kladderadeng** [Ylianst/MeshCentral#6491]
**Describe the bug**
When remove accessing another System and for example using CTRL+C or CTRL+V for Copy & Paste, most of the time a c or v is written. The Key combination only works for me, when I wait half a second or a second after depressing the CTRL Key until I press C or V. When using Copy & Paste over a Remote session often, this makes it very cumbersome to use. I feel this is more an issue in the Webbrowser than when using the Router Windows executable. But the latter one seems to miss the feature to transfer the Clipboard from the Remote to the local System, so it is less preferred.

**To Reproduce**
1. Remote access another Windows 11 24H2 System either through the Browser or through the Windows Router executable.
2. Select any kind of text on the Screen (for example in Notepad)
3. Press CTRL+C or CTRL+V as you probably will when working locally (pretty short key presses)
4. You will see a c or a v on the screen and whatever Text you selected will be substituted by that single character instead of copying it into / pasting it from the clipboard.

**Expected behavior**
I would expect that keys that are pressed in a specific order are sent to the destination in that order. But this does not seem to be the case when CTRL is involved.

**Server Software (please complete the following information):**
 - OS: Ubuntu Server 22.04.5 LTS
 - Virtualization: Docker Image ghcr.io/ylianst/meshcentral:1.1.32
 - Network: Both testsystems are on the same Network (LAN), but Meshcentral runs through Traefik and I am using Hybrid mode
 - Version: 1.1.32
 - Node: 20.15.1

**Client Device (please complete the following information):**
 - Device: PC
 - OS: Windows 11 23H2
 - Network: LAN (Source PC) to VLAN (Traefik + Meshcentral) to LAN (Destination PC)
 - Browser: Microsoft Edge and Vivaldi
 - MeshCentralRouter Version: 1.8.9021

**Remote Device (please complete the following information):**
 - Device: VM under XCP-ng
 - OS: Windows 11 24H2
 - Network: LAN
 - Current Core Version: Dec 9 2022, 1874374695

**Additional context**

**Your config.json file**
```
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "settings": {
    "Cert": "meshcentral.my.domain",
    "plugins":{"enabled": false},
    "_mongoDb": null,
    "cert": "localhost",
    "_WANonly": true,
    "_LANonly": true,
    "_sessionKey": "redacted",
    "port": 443,
    "aliasPort": 443,
    "redirPort": 80,
    "redirAliasPort": 80,
    "AgentPong": 300,
    "TLSOffload": true,
    "SelfUpdate": false,
    "AllowFraming": false,
    "Minify": 1,
    "WebRTC": true
  },
  "domains": {
    "": {
      "title": "MyLab",
      "title2": "Meshcentral",
      "minify": true,
      "NewAccounts": false,
      "localSessionRecording": false,
      "_userNameIsEmail": true,
      "certUrl": "meshcentral.my.domain",
      "allowedOrigin": [ "meshcentral.my.domain" ]
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before>",
    "_email": "myemail@mydomain.com",
    "_names": "myserver.mydomain.com",
    "production": false
  }
}
```

**Comment by si458** [comment]
Have u tried using the 'use remote keyboard map' option under settings?

Sometimes, this is needed, and sometimes, it isn't

For example when i control my mac, i need to untick it, but when I load up utm on my remote mac and then try typing inside the vm window, it won't type, so I have to tick the 'use remote keyboard map' then I can type inside the vm window

**Comment by kladderadeng** [comment]
Thank you for your reply. I just fired up the test environment again and this option does not help. I do not have problems with typing in general. I touch type decently fast and when I do that in a remote session, it all seems to work fine. Even Upercase / Lowercase, where SHIFT is used as a modifier. The issue seems to be specific about the CTRL Key. I suspect this is handled. Even the ALT key seems to work fine... but with CTRL I have to wait .5 - 1 second until it does its thing. I would not even open the issue if the Router had more options (like bi-directional Clipboard support like we have in the browser) and would just stick to the Router ;-)

**Comment by si458** [comment]
The meshcentralrouter has clipboard built in if I'm correct, u don't need to enable/disable any options?

Also have u tried going browser fullscreen?
Use edge/chrome, connect to device, then HOLD SHIFT while clicking the fullscreen icon,
This enabled the browser to take full control of all key presses, so it might help, maybe?

**Comment by kladderadeng** [comment]
Yes, you are right. The meshcentralrouter has clipboard built in. I made the mistake and activated the "Automatically send clipboard" feature in the router, which automates the clipboard but only in one direction. When I turn that off, I see the 2 clipboard Icons for manual transfer. Thank you!

I just tried browser fullscreen with SHIFT + Click on the fullscreen icon. This did not change the CTRL behavior. 
I also just found out that ALT-GR is not working either in the Browser, but that is not a big issue because you can emulate it with CTRL+ALT (requires also the waiting time for CTRL ;-)).

I did not find the other CTRL issue, I am sorry. But I can confirm that it is not sticky. It just takes the 0,5 to 1 second to "activate".

Update! It only seems to affect the LEFT CTRL Key. When I use the right one, it works instantly like I would expect. I never tried, because left CTRL+V / C need only 1 hand to activate.
Also I went to www.keyboardtester.com on the Remote machine to see which keys activate when. When I press for example SHIFT 5 times in quick succession, all 5 key presses are registered. If I do this with CTRL, depending how quick i am I get 0-2 registered. CTRL seems to be a special snowflake here ;-)
I recorded a short video about this, but I do not know how helpful this is.

https://github.com/user-attachments/assets/721d47f2-998e-4569-9335-2e5c14a99317

**Comment by si458** [comment]
> I made the mistake and activated the "Automatically send clipboard" feature in the router, which automates the clipboard but only in one direction

erm thats not right if i remember it should be both ways???
i would open an issue in the [meshcentralrouter](https://github.com/Ylianst/MeshCentralRouter) repo about that
and we can look into it but at least u can use the manual buttons!

as for the CTRL key issue,
sounds mad but you screenrecording shows you use webrtc,
try disabling it in config.json `webrtc: false`
restarting meshcentral and see if the issue still happens?

**Comment by kladderadeng** [comment]
Thank you for your comment. I just tried with webrtc: false and a restart, still the same issue. Left CTRL is delayed, right CTRL is fine. Is there some kind of "communication log" that logs all the keypresses that are sent back and forth that might help here?

I will double check the clipboard issue in router and then go to the other Repo ;-)

**Comment by si458** [comment]
you can use `?trace=1` in your address bar to enable to websocket debug
and you can also use `?redirtrace=1` to debug the tunnel traffic that the mouse/keyboard uses
or use them both together at the same time `?trace-1&redirtrace=1`
then check ur developer tools of the browser and watch the console.log fly by haha

**Comment by kladderadeng** [comment]
Oh yea. That worked! Thank you very much. As long as you do not move the mouse, that information is quite readable / traceable ;-)

Here are the results:
![image](https://github.com/user-attachments/assets/6a138c16-f9c0-4aa5-a435-d64704a78977)

Green: Left Shift Depress and Release
Red: Left CTRL Depress and Release (only Depress is transferred)
Purple: Right CTRL Depress and Release

When I hold the left CTRL longer, Depress and Release is recorded.

Here I try a CTRL+C like you would normally do it, pretty quickly (first CTRL, then C of course).
043 is depressing C
111 is depressing CTRL (I pressed that BEFORE the C, I am 100% sure!)
143 is releasing C
There is no releasing of CTRL recorded.

![image](https://github.com/user-attachments/assets/2c4916f4-6ad9-461f-aee5-e34e00b7edda)


Here I try a CTRL+C while pressing CTRL, wait a bit and then pressing C:
011 is depressing CTRL (I waited a bit before pressing the next key)
043 is depressing C
143 is releasing C
111 is releasing CTRL
![image](https://github.com/user-attachments/assets/d037d9f8-26dd-449c-bcab-b6bb9ddcfdc9)


I am not sure how this is helpful, but to me it looks like it takes longer for the program to send the CTRL through the connection and when I do a CTRL+C quickly, c is sent before the CTRL and of course also arrives before the CTRL on the remote system. The only thing that this test is telling me at the moment is, that this should not be an issue on the remote side. If what I see in the Debug logs is what leaves my PC, my side fails to send a CTRL as quickly as the other keys and this messes up the order in which I am typing.

**Comment by Chris2381231** [comment]
Just adding on to this as we have the same issue - only Left CTRL is affected.

I have observed this issue on several devices but can't confirm any specific OSEs besides Windows 10 Pro having the issue.

Current version (same as kladderadeng)
- 1.1.32

Previously versions used (did not present the issue)
- 1.1.21
- <= 1.1.5

Made a quick gif below to demonstrate.
- Top On-Screen Keyboard = Remote Session
- Bottom On-Screen Keyboard = Local Device

The initial CTRL key presses are with Left CTRL. None of the key presses are registered besides the last key press.
All Right CTRL key presses are registered, regardless of how fast the key presses are.

![ss (2024-12-27 at 13 02 44)](https://github.com/user-attachments/assets/2b1feecc-594c-4933-b13f-385fbdb6fac9)

**Comment by geekasso** [comment]
I can confirm that I have the same issue. I tried with webrtc set to false. If I use CTRL+A, for example, I have to press the CTRL for a half a second before the A.

**Comment by redddcyclone** [comment]
I'm having this same problem. Only happens with LeftCtrl. Windows 11 24H2. Doesn't matter if WebRTC is on or off.

**Comment by bipton** [comment]
I have the same issue with the ctrl key, I use ctrl-c or ctrl-v whether it's a windows client or linux client and it's runs off with multiple c or v's. I am running server 1.1.46.

**Comment by cy11100** [comment]
I've had this issue consistently over multiple versions. It's gotten to the point that I've started using right click context menu to copy and paste when using Meshcentral. 

I've seen and tried the suggested solutions in previous cases where I've searched for solutions, such as https://github.com/Ylianst/MeshCentral/issues/5739 (disable webrtc, wscompression, enable remote keyboard map) without any success. I've actually just checked my `config.json` and these options were still active from my last attempt at addressing this bug.

Good on @kladderadeng for figuring out it's a delay of the ctrl key. I can verify that I am experiencing the same issue with ctrl key being delayed, and verified that holding ctrl down for 1 second executes the command correctly. I can also verify that the issue only applies for the left ctrl key, like others have mentioned.

**Comment by Chris2381231** [comment]
> you can use `?trace=1` in your address bar to enable to websocket debug and you can also use `?redirtrace=1` to debug the tunnel traffic that the mouse/keyboard uses or use them both together at the same time `?trace-1&redirtrace=1` then check ur developer tools of the browser and watch the console.log fly by haha

I think I've found the smoking gun but not sure what a proper solution would be without breaking AltGr again #5614.

https://github.com/Ylianst/MeshCentral/commit/6e4c8b0d164dd997b48fdbd32ad1be9532e4c360
- Pushed on version 1.1.17 (my previous comment mentioned 1.1.21 seemed unaffected, although we were only on that version for a month)

Left CTRL presses within ~100-200ms are not registering, which lines up with the handleAltGrTimeout of 100ms.
- AltGr sequence is only checking for ControlLeftKc, which is why Right CTRL is unaffected

I assume reducing the timeout would have its own set of issues. Not sure if it would be possible to restructure the approach to not have a timeout at all, or introduce a toggle to turn it off / on.

x2 Left CTRL > x2 Right CTRL (quick presses)
![Image](https://github.com/user-attachments/assets/8effc1fb-04a7-4cf6-a1ca-68af72367c6a)

x2 Left CTRL > x2 Right CTRL (slow presses)
![Image](https://github.com/user-attachments/assets/87f57831-d1a2-4585-9da3-ac429aaf23fd)

**Comment by aelfwine88** [comment]
This still an annoying issue, please don't close. 🙃

**Comment by redddcyclone** [comment]
Will this issue ever get an attention? It's terribly annoying, it affects the use of Left CTRL *a lot*...

**Comment by si458** [comment]
@redddcyclone please remember this is a community project.
The are no full time developers and we are all working in our free time.
We will have a look when we get chance 
Me for example im in the middle of adding features and fixing other issues people are having with the web ui

**Comment by mark-pippin** [comment]
# This is the work of an LLM (fable 5), caveats apply. I don't have time to take sentient ownership.

# Root cause found + working self-apply patch (tested on v1.2.0)

I hit this bug, traced it to its root cause, patched it locally, and have it verified working. Since the issue keeps auto-closing, here's a full writeup: what the bug is, a patch users can self-apply **without forking or rebuilding**, and enough detail for a maintainer (or anyone) to turn it into a proper PR.

## TL;DR

`checkAltGr()` in `public/scripts/agent-desktop-0.0.2.js` (added in 6e4c8b0 / v1.1.17 for #5614) defers every **Left-Ctrl keydown** on Windows browsers for up to 100 ms, waiting to see whether an AltRight follows (Windows signals AltGr as a fake Ctrl+Alt pair). Two defects in that code:

1. **The deferred Ctrl-down is dropped, not flushed.** If *any other key* arrives inside the 100 ms window, the disarm path cancels the timer and simply never sends the Ctrl keydown. A fast Ctrl+C reaches the target as `C` alone; a fast Ctrl tap never reaches the target at all.
2. **A broken membership test:** `ControlLeftKc in obj.pressedKeys` — `pressedKeys` is an **array**, so `in` tests *indices* (is there an element at index 17?), never membership. The guard never does what it intends.

This exactly matches every observation in this thread:

- **Only Left Ctrl** — only `event.code === "ControlLeft"` enters this path; Right Ctrl is untouched.
- **Hold ~0.5–1 s and it works** — the 100 ms timeout fires and finally sends the Ctrl-down; after that the combo works.
- **Rapid Ctrl taps register 0–2 of 5** (@kladderadeng's keyboardtester test) — any tap shorter than 100 ms is swallowed; only taps spaced longer than the timeout get delivered.
- **@kladderadeng's `?redirtrace=1` capture** (`043`, `111`, `143` on a fast Ctrl+C): decoding against his slow capture (`011` = Ctrl down, `111` = Ctrl **up**), the fast trace is actually `C↓, Ctrl↑, C↑` — the Ctrl **down** was swallowed by the disarm path, while the Ctrl **up** leaked through via the (broken) re-arm branch, which returns `false` for keyups. The trace is a perfect fingerprint of defect 1.
- **Started in v1.1.17** — matches @Chris2381231's version bisection and the 6e4c8b0 commit he identified. Still present in v1.2.0.

## The patch

Against v1.2.0 `public/scripts/agent-desktop-0.0.2.js`. Design goal: **zero added latency, zero lost keys, AltGr merging preserved unchanged** (both the keydown merge and the keyup merge). The idea: when the pending Left-Ctrl turns out *not* to be the start of an AltGr sequence, **flush the deferred Ctrl-down first**, then process the interrupting key — instead of dropping it.

```diff
--- a/public/scripts/agent-desktop-0.0.2.js
+++ b/public/scripts/agent-desktop-0.0.2.js
@@ (object state, ~line 44)
     obj._altGrArmed = false;       // Windows AltGr detection
     obj._altGrTimeout = 0;
+    obj._altGrDeferredDown = false; // a ControlLeft keydown is held back pending AltGr detection
@@ (checkAltGr, ~line 446)
         if (obj._altGrArmed) {
             obj._altGrArmed = false;
             clearTimeout(obj._altGrTimeout);
+            var deferredCtrlDown = obj._altGrDeferredDown;
+            obj._altGrDeferredDown = false;
 
             if ((event.code === "AltRight") &&  ((event.timeStamp - obj._altGrCtrlTime) < 50)) {
                 //AltGr detected.
                 obj.SendKeyMsgKC( action, AltGrKc, false);
                 return true;
             }
+            // Not an AltGr sequence: flush the deferred ControlLeft keydown BEFORE processing
+            // this event, so the remote sees Ctrl go down first. Previously it was dropped
+            // here, which lost fast Left-Ctrl taps and delivered "C before Ctrl" on a quick
+            // Ctrl+C (issue #6491).
+            if (deferredCtrlDown) { obj.SendKeyMsgKC( 1, ControlLeftKc, false); }
         }
 
         // Possible start of AltGr sequence?
-        if ((event.code === "ControlLeft") && !(ControlLeftKc in obj.pressedKeys)) {
+        // NOTE: `in` on an array tests INDICES, not membership — must be indexOf. Arming on
+        // keyup (no timer, falls through to return false so the keyup still sends) is
+        // intentional: it lets the AltRight keyup that follows merge into an AltGr keyup.
+        if ((event.code === "ControlLeft") && (obj.pressedKeys.indexOf(ControlLeftKc) == -1)) {
           obj._altGrArmed = true;
             obj._altGrCtrlTime = event.timeStamp;
           if( action == 1 ) {
+            obj._altGrDeferredDown = true;
             obj._altGrTimeout = setTimeout(obj._handleAltGrTimeout.bind(obj), 100);
             return true;
           }
         }
         return false;
     }
 
     obj._handleAltGrTimeout = function () { //Windows and no Ctrl+Alt -> send only Ctrl.
         obj._altGrArmed = false;
+        obj._altGrDeferredDown = false;
         clearTimeout(obj._altGrTimeout);
         obj.SendKeyMsgKC( 1, ControlLeftKc, false); // (KeyDown, "ControlLeft", false)
     }
```

### Behavior after the patch (traced event-by-event)

| Scenario | Before | After |
|---|---|---|
| Fast Ctrl+C (< 100 ms) | remote sees `C↓ Ctrl↑ C↑` — Ctrl-down dropped | remote sees `Ctrl↓ C↓ C↑ Ctrl↑` |
| Quick lone Ctrl tap | nothing (down dropped, up orphaned) | `Ctrl↓ Ctrl↑` |
| Ctrl held > 100 ms, then key | works (timeout fires) — unchanged | works — unchanged |
| Real AltGr (Ctrl↓+AltRight↓ < 50 ms apart) | merged to `AltGr↓` | merged to `AltGr↓` — unchanged |
| AltGr release (Ctrl↑+AltRight↑) | merged to `AltGr↑` | merged to `AltGr↑` — unchanged |
| AltRight arriving ≥ 50 ms after Ctrl↓ | Ctrl-down dropped | Ctrl-down flushed, then AltRight processed |

The only remaining artifact of the heuristic is that a *lone* Left-Ctrl press still reaches the target ~100 ms late — harmless, and inherent to AltGr detection by timeout (Windows browsers genuinely give no other way to distinguish Left-Ctrl from the start of an AltGr pair).

## Self-apply without forking (survives server updates)

Use MeshCentral's supported **`meshcentral-web` override tree** — files under `meshcentral-web/public/...` shadow the installed `public/...` (checked first by the static file router). No rebuild, no fork:

1. Locate your install. Docker: the `meshcentral-web` folder sits beside `meshcentral-data` in the container (`/opt/meshcentral/meshcentral-web`) — bind-mount it (e.g. `- ./web:/opt/meshcentral/meshcentral-web`). Bare install: create `meshcentral-web/` as a **sibling of `meshcentral-data/`**.
2. `mkdir -p meshcentral-web/public/scripts`
3. Copy the **pristine file from your installed version** (e.g. `node_modules/meshcentral/public/scripts/agent-desktop-0.0.2.js`, or `docker cp` it out) into that folder and apply the diff above.
4. **Also copy the patched file to `agent-desktop-0.0.2-min.js` in the same folder.** Which name is served depends on `settings.minify` in config.json; covering both names covers both configurations (serving the unminified content under the `-min` name is functionally fine).
5. Restart MeshCentral, then **hard-refresh the browser (Ctrl-F5)** — the old script is cached.

Verify: open a remote session with `?trace=1&redirtrace=1` and do a fast Ctrl+C — you should now see Ctrl-down (`011`) *before* C-down (`043`).

Caveat: an override shadows the *whole file*, so after upgrading MeshCentral re-diff your override against the new version's copy and re-apply.

## Results

Tested on MeshCentral v1.2.0 (docker, `ghcr.io/ylianst/meshcentral`), Chrome on Windows → Windows 11 target:

- Fast Ctrl+C / Ctrl+V / Ctrl+A at normal typing speed: register correctly, no hold-and-wait needed.
- Rapid repeated Left-Ctrl taps: all register on the target (previously 0–2 of 5).
- Right Ctrl: unchanged (was never affected).
- Verified in both the native web UI and an iframe-embedded desktop session.

**Not personally re-tested:** AltGr on an international layout (I'm on a US layout). The merge logic for the detected-AltGr path — the reason this code exists (#5614) — is deliberately untouched by the patch, and I've traced the keydown *and* keyup merge sequences as preserved, but a test from someone with a German/Nordic layout against a Linux target would be a valuable confirmation.

## Notes for whoever turns this into a PR

- Canonical source: `public/scripts/agent-desktop-0.0.2.js` (`checkAltGr` + `_handleAltGrTimeout`; state init near the top of `CreateAgentRemoteDesktop`). The committed sibling `agent-desktop-0.0.2-min.js` must be regenerated (or receive the equivalent edit) — it's what most installs serve, since `minify` defaults on.
- `grep -r checkAltGr` to confirm there are no other copies; the mobile UI loads the same script. MeshCentral Router is a separate native codebase and doesn't have this bug (consistent with reports in this thread that Router behaves correctly).
- Suggested test matrix: fast Ctrl+C/V/A; 5 rapid Ctrl taps; Ctrl held through multiple keys; AltGr characters on an intl layout from a Windows browser against **Windows and Linux** targets (the #5614 case); AltGr held with autorepeat.
- The `in` → `indexOf` fix is worth keeping even independently of the flush: as written, the guard can re-arm on autorepeated Ctrl keydowns and on keyups in ways the author almost certainly didn't intend (it happens to make the AltGr-keyup merge work by accident — the patch keeps that behavior deliberately and documents it).
- Alternative design if maintainers ever want *zero* latency on lone Ctrl as well: send Ctrl-down immediately (no deferral), and when an AltRight arrives < 50 ms later, send Ctrl-up then AltGr-down (the transient remote Ctrl press is harmless on all targets since bare Ctrl has no side effect). That removes the timeout entirely, at the cost of a slightly noisier event stream on the target during AltGr. The flush approach above was chosen because it's the minimal change to the existing structure.

**Comment by si458** [comment]
@mark-pippin happy for you to ask it to submit a PR!
we can then check it over fully!

**Comment by redddcyclone** [comment]
Hi @si458

I tested the changes @mark-pippin posted and can confirm that this fixes this issue. I posted PR #8040 so that you can review it.
I left the original comments just because, but I can remove them if needed.
